example/src/Main.purs

```diff
@@ -0,0 +1,327 @@
+          [ DurationF.Year
+          , DurationF.Month
+          , DurationF.Day
+          , DurationF.Hour
+          , DurationF.Minute
+          , DurationF.Second
+          ]
+          "YYYY:MM:DD-HH:mm"
+        )
+        (Right $ DurationEnd testDuration testDateTime)
+    <> renderInterval s 3
+        (StartEnd "YYYY:MM:DD" "YYYY:MM:DD")
+        (Right $ StartEnd testDateTime testDateTime)
+
+  testDate :: Date
+  testDate = canonicalDate (enum 2017) (enum 1) (enum 1)
+
+  testDateTime :: DateTime
+  testDateTime = DateTime testDate (bottom # setHour (enum 2) # setMinute (enum 2))
+
+  testDuration :: IsoDuration
+  testDuration = unsafePartialBecause "this duration must be valid" fromJust $ I.mkIsoDuration $ I.year 100.0 <> I.month 25.0 <> I.day 245.0 <> I.hour 0.0 <> I.minute 100.0 <> I.second 124.0
+
+  enum :: ∀ a. BoundedEnum a => Int -> a
+  enum = unsafePartialBecause "ints passed to this func must be in range" fromJust <<< toEnum
+
+  renderTime ∷ State -> Int -> String -> Either String Time -> Array (HTML m)
+  renderTime s = renderExample timeConfig s.times
+
+  renderDate ∷ State -> Int -> String -> Either String Date -> Array (HTML m)
+  renderDate s = renderExample dateConfig s.dates
+
+  renderDuration ∷ State -> Int -> Array DurationF.Command -> Either String IsoDuration -> Array (HTML m)
+  renderDuration s = renderExample durationConfig s.durations
+
+  renderInterval ∷ State -> Int -> Interval (Array DurationF.Command) String -> Either String (Interval IsoDuration DateTime) -> Array (HTML m)
+  renderInterval s = renderExample intervalConfig s.intervals
+
+  renderDateTime ∷ State -> Int -> String -> Either String DateTime -> Array (HTML m)
+  renderDateTime s = renderExample dateTimeConfig s.dateTimes
+
+  eval ∷ Query ~> DSL m
+  eval (Set payload next) = do
+    map mustBeMounted $ case payload of
+      SetTime     idx val -> H.query' timeConfig.cp     idx $ H.request $ left <<< Base <<< (SetValue $ map Right val)
+      SetDate     idx val -> H.query' dateConfig.cp     idx $ H.request $ left <<< Base <<< (SetValue $ map Right val)
+      SetDateTime idx val -> H.query' dateTimeConfig.cp idx $ H.request $ left <<< Base <<< (SetValue $ map Right val)
+      SetDuration idx val -> H.query' durationConfig.cp idx $ H.request $ left <<< Base <<< (SetValue $ map Right val)
+      SetInterval idx val -> do
+        res <- H.query' intervalConfig.cp idx $ H.request $ left <<< Base <<< (SetValue $ map Right val)
+        pure $ void res -- <#> (\error ->  D.traceAny {message:"can't update interval", error})
+    pure next
+  eval (HandleMessage payload next) = do
+    case payload of
+      MsgTime     idx (NotifyChange val) -> H.modify \s -> s {times     = insert idx (show val) s.times}
+      MsgDate     idx (NotifyChange val) -> H.modify \s -> s {dates     = insert idx (show val) s.dates}
+      MsgDateTime idx (NotifyChange val) -> H.modify \s -> s {dateTimes = insert idx (show val) s.dateTimes}
```

Comment:
✅

didn't get what you mean by "abstract things like insert idx (show val) and it" tho